### PR TITLE
[Messenger] add missing log context when sending to failure transport

### PR DESCRIPTION
**Pull request by ajardin**
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | N/A

## Why?
Logs generated by `SendFailedMessageToFailureTransportListener::onMessageFailed` do not include the message in the context, unlike those generated by `SendFailedMessageForRetryListener::onMessageFailed`.

## Before
```
INFO      [messenger] Rejected message App\Message\MyCustomMessage will be sent to the failure transport Symfony\Component\Messenger\Bridge\Doctrine\Transport\DoctrineTransport.
 [
   "class" => "App\Message\MyCustomMessage",
   "transport" => "Symfony\Component\Messenger\Bridge\Doctrine\Transport\DoctrineTransport"
 ]
 [
   "request_id" => false,
   "uid" => "537143b"
 ]
```

## After
```
INFO      [messenger] Rejected message App\Message\MyCustomMessage will be sent to the failure transport Symfony\Component\Messenger\Bridge\Doctrine\Transport\DoctrineTransport.
 [
   "message" => App\Message\MyCustomMessage {
     -foo: true
     -bar: false
   },
   "class" => "App\Message\MyCustomMessage",
   "transport" => "Symfony\Component\Messenger\Bridge\Doctrine\Transport\DoctrineTransport"
 ]
 [
   "request_id" => false,
   "uid" => "537143b"
 ]
```

### CODE REVIEW

**Comment by chalasr** [comment]
Thanks for the PR. This change looks like a new feature to me, or is there something that breaks without it?

**Comment by ajardin** [comment]
Hello and thanks for the reply! 🙂 

You're right, it's not _broken_. I was also hesitating between a bug and a feature...
Basically, it makes the log less useful as the relationship between the message and the event is not visible, which makes me think it's a _missing_ element rather than a _new_. What do you think?

**Comment by fabpot** [comment]
That’s a new feature 😀

**Comment by ajardin** [comment]
I have updated the pull request accordingly. 😉

**Comment by lyrixx** [comment]
Message content can be very huge. This can also overflow some third party service.

But I like the idea. I'm my application, I usually log raw object, then I add a custom monolog processor to keep only what I need. (I gave a [talk](https://speakerdeck.com/lyrixx/symfony-live-2015-paris-monitorer-sa-prod?slide=38) about it)

We could do the same here

1. Log with the raw envelop
2. Add a default monolog processor that
    *  Remove eveything and keep only the class.
    * **Or** keep only the 1024 first bytes of the content (arbitrary number)

So if people want more or less information in the log, they can decorate / replace the default processor in the DIC.

---

WDYT?

(I actually really like the idea. I think it could be generalized to the framework, when applicable)

**Comment by ajardin** [comment]
Your idea goes further than my initial thoughts, but I really like it! 🙂 

Instead of adding an additional risk of overflow (the message is already logged in multiple places), we would fix a minor lack and improve the component monitoring. Moreover, the possibility to replace/decorate the processor would be great.

I can try to work on this approach in the coming days if nobody else is on it.
In the meantime, what do you want to do with this pull request?

**Comment by lyrixx** [comment]
> I can try to work on this approach in the coming days if nobody else is on it.

Please, wait for more approval from the core team, I'm not sure everything is OK with that

ping @symfony/mergers

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php
Comment (Nyholm) [inline src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php:65]: I suggest ro remove this key.

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php
Comment (Nyholm) [inline src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php:66]:
```suggestion
            'message_class' => \get_class($message),
```

**Review by Nyholm — CHANGES_REQUESTED** [review]
I agree with @lyrixx. This is a great feature. But messages can be huge. Messages can also include sensitive data (passwords) that should not end up in the logs. 

I suggest to remove the `message`. Adding the message class is little but enough info to add more logging in your app if you need to know the exact contents of the message.

**Comment by ajardin** [comment]
@Nyholm removing this key is basically closing the pull request. 😅 

Joke aside, what is your opinion on the processor approach proposed by @lyrixx? We could only keep the message class by default and let users decide if they want to add further data through decorators.

Because even if we decide to not add the message within `SendFailedMessageToFailureTransportListener::onMessageFailed` logs, the message is already available (at least) in:
- `\Symfony\Component\Messenger\TraceableMessageBus::dispatch`
- `\Symfony\Component\Messenger\Worker::ack`
- `\Symfony\Component\Messenger\EventListener\SendFailedMessageForRetryListener::onMessageFailed`
- `\Symfony\Component\Messenger\Middleware\HandleMessageMiddleware::handle`
- `\Symfony\Component\Messenger\Middleware\SendMessageMiddleware::handle`

**Comment by ajardin** [comment]
Hello @lyrixx and @Nyholm,

Did you have the opportunity to discuss the changes mentioned above? Thanks for your time.

**Comment by lyrixx** [comment]
Adding the raw message is a no go unfortunately. IMHO the custom processor is the way to go.

**Comment by ajardin** [comment]
Thanks for your feedback. In that case, I'm closing this pull request. 

But while I understand the reason why adding the message could cause problems (e.g. leak of sensitive data or logs overflow), I don't understand why it's a problem here and not in the other parts of the component. Am I missing something?

**Comment by Nyholm** [comment]
You are correct. Maybe we did miss something in previous prs. 

I would be super happy if you want to help by removing those instances to. 
I definitely understand if you don’t, but please tell us so someone else can work on that.

**Comment by ajardin** [comment]
No problem, I'll submit a PR with the removal of other instances.

But I would love to have a workaround idea for the application I'm working on since we use the message details to monitor exactly what's going on. 😉 

Should we subscribe to Messenger events and add our own log entries?

**Comment by Nyholm** [comment]
> Should we subscribe to Messenger events and add our own log entries?

Yes, 
You can also add your custom middleware. That will give you maximum control.

**Pull request closed without merging**
